# Ejercicio ETL

### Resumen

A partir de una fuente de datos opensource en [Kaggle](https://www.kaggle.com/datasets/michau96/restaurant-business-rankings-2020) hemos creado montado ETL. Extraído, procesado y cargado los datos en una BBDD (base de datos).


---


## Obtención de datos

INPUT: 3 archivos CSV con las clasificaciones de los mejores restaurantes de 2020.



*   Future 50:
Clasificaciones de los futuros 50 mejores restaurantes.
*   Independence 100:
Clasificaciones de los 100 mejores restaurantes independientes de 2020.
*   Top 250:
Clasificaciones de los 250 mejores restaurantes de 2020.

Todos los datos se extrajeron de Restaurant Business Online y están basados en informes financieros de la empresa, encuestas, documentos de las franquicias y algoritmos de valoración.

In [2]:
#Dependencias
import pandas as pd
import numpy as np
#Instalamos e importamos MySQL para Python
!pip install PyMySQL --quiet
import pymysql as pymsql
!pip install mysql-connector --quiet
import mysql.connector
from sqlalchemy import create_engine
!pip install sqlalchemy-utils --quiet
from sqlalchemy_utils import database_exists, create_database

     |████████████████████████████████| 43 kB 1.7 MB/s 
     |████████████████████████████████| 11.9 MB 7.6 MB/s 
     |████████████████████████████████| 100 kB 4.5 MB/s 


## Extraer CSVs en DataFrame

In [6]:
#1. Leer el archivo top250 y ver el dataframe
top250_file = "https://drive.google.com/uc?id=10fsqUk01wJeLUcArcBN2JUg86C3lKLQm&export=download"
top250 = pd.read_csv(top250_file)
top250

Rank           Restaurant  \
0       1           McDonald's   
1       2            Starbucks   
2       3          Chick-fil-A   
3       4            Taco Bell   
4       5          Burger King   
..    ...                  ...   
245   246      Rainforest Cafe   
246   247                  PDQ   
247   248        Lupe Tortilla   
248   249  Cook-Out Restaurant   
249   250             Jollibee   

                                               Content  Sales YOY_Sales  \
0                                                  NaN  40412      4.9%   
1                                                  NaN  21380      8.6%   
2    While Popeyes got a lot of the chicken buzz in...  11320     13.0%   
3                                                  NaN  11293      9.0%   
4                                                  NaN  10204      2.7%   
..                                                 ...    ...       ...   
245                                                NaN    129    -10.4%   
246                                                NaN    127     -5.5%   
247                                                NaN    127     12.1%   
248                                                NaN    126     10.1%   
249                                                NaN    126     15.2%   

     Units YOY_Units Headquarters             Segment_Category  
0    13846     -0.5%          NaN       Quick Service & Burger  
1    15049      3.0%          NaN  Quick Service & Coffee Cafe  
2     2470      5.0%          NaN      Quick Service & Chicken  
3     6766      2.7%          NaN      Quick Service & Mexican  
4     7346      0.2%          NaN       Quick Service & Burger  
..     ...       ...          ...                          ...  
245     18     -5.3%          NaN                  Varied Menu  
246     56    -11.1%          NaN                      Chicken  
247     25      8.7%          NaN                      Mexican  
248    270      7.1%          NaN                       Burger  
249     40     11.1%          NaN                      Chicken  

[250 rows x 9 columns]

In [9]:
#2. Leer el archivo Future50 y ver el dataframe
future50_file = "https://drive.google.com/uc?id=15zBrVc3kx7QxYkNaFaNVfh-a0uj-TKbo&export=download"
#Carga el archivo future50_file en el dataframe future50
future50 = pd.read_csv(future50_file)
future50

Rank                        Restaurant                  Location  Sales  \
0      1                        Evergreens            Seattle, Wash.     24   
1      2                       Clean Juice           Charlotte, N.C.     44   
2      3                          Slapfish  Huntington Beach, Calif.     21   
3      4                        Clean Eatz          Wilmington, N.C.     25   
4      5                         Pokeworks            Irvine, Calif.     49   
5      6                       Playa Bowls             Belmar,  N.J.     39   
6      7                  The Simple Greek            Blue Bell, Pa.     24   
7      8                         Melt Shop            New York, N.Y.     20   
8      9                        Creamistry      Yorba Linda,  Calif.     24   
9     10              Joella's Hot Chicken           Louisville, Ky.     29   
10    11                     Eggs Up Grill         Spartanburg, S.C.     30   
11    12                          Dog Haus          Pasadena, Calif.     39   
12    13                  Teriyaki Madness             Denver, Colo.     41   
13    14                    Bluestone Lane            New York, N.Y.     48   
14    15                 Original ChopShop              Plano, Texas     21   
15    16                 Rapid Fired Pizza           Kettering, Ohio     24   
16    17           Ike's Love & Sandwiches     San Francisco, Calif.     44   
17    18                    Vitality Bowls         San Ramon, Calif.     37   
18    19         Hawkers Asian Street Fare             Orlando, Fla.     22   
19    20          Maple Street Biscuit Co.         Orange Park, Fla.     39   
20    21                   Bulla Gastrobar               Doral, Fla.     32   
21    22                       Duck Donuts        Mechanicsburg, Pa.     44   
22    23                   The Little Beet            New York, N.Y.     23   
23    24                   Joe & The Juice            New York, N.Y.     47   
24    25                          By Chloe            New York, N.Y.     37   
25    26              Sugarfire Smokehouse             Olivette, Mo.     39   
26    27               Bibibop Asian Grill            Columbus, Ohio     44   
27    28               Bubbakoo's Burritos       Wall Township, N.J.     28   
28    29                Dos Toros Taqueria            New York, N.Y.     28   
29    30                   The Pizza Press           Anaheim, Calif.     27   
30    31                    The Lost Cajun             Frisco, Texas     20   
31    32                        Just Salad            New York, N.Y.     42   
32    33        Jeni's Splendid Ice Creams            Columbus, Ohio     42   
33    34                          Boqueria            New York, N.Y.     27   
34    35         Quickway Japanese Hibachi              Fairfax, Va.     22   
35    36                         Surcheros              Douglas, Ga.     22   
36    37                   Famous Toastery           Charlotte, N.C.     40   
37    38                  Culinary Dropout         Scottsdale, Ariz.     20   
38    39                     Condado Tacos            Columbus, Ohio     38   
39    40           The Flying Biscuit Cafe              Atlanta, Ga.     29   
40    41             Blue Sushi Sake Grill               Omaha, Neb.     49   
41    42                    The Human Bean             Medford, Ore.     47   
42    43   Spitz Mediterranean Street Food       Los Angeles, Calif.     28   
43    44                      Tacos 4 Life              Conway, Ark.     25   
44    45    Pita Mediterranean Street Food             Fairburn, Ga.     38   
45    46                       LA Crawfish            McAllen, Texas     48   
46    47                            &pizza          Washington, D.C.     45   
47    48               Super Duper Burgers     San Francisco, Calif.     39   
48    49                   StoneFire Grill      Agoura Hills, Calif.     31   
49    50  Gus's World Famous Fried Chicken            

In [20]:
#3. Leer el archivo Independence100 y ver el dataframe
ind100_file = "https://drive.google.com/uc?id=1cg8-zWjWkD1iRuRMDs5DZmSqnhHhHNcA&export=download"
#Carga el archivo ind100_file en el dataframe ind100
ind100 = pd.read_csv(ind100_file)
ind100

Rank                           Restaurant       Sales  Average Check  \
0      1             Carmine's (Times Square)  39080335.0             40   
1      2                The Boathouse Orlando  35218364.0             43   
2      3                     Old Ebbitt Grill  29104017.0             33   
3      4  LAVO Italian Restaurant & Nightclub  26916180.0             90   
4      5             Bryant Park Grill & Cafe  26900000.0             62   
..   ...                                  ...         ...            ...   
95    96                 George's at the Cove  12194000.0             80   
96    97                            Le Coucou  12187523.0             95   
97    98                              Mi Vida  12032014.0             38   
98    99                               Upland  11965564.0             52   
99   100               Virgil's Real Barbecue  11391678.0             27   

          City   State  Meals Served  
0     New York    N.Y.      469803.0  
1     Orlando     Fla.      820819.0  
2   Washington    D.C.      892830.0  
3     New York    N.Y.      198500.0  
4     New York    N.Y.      403000.0  
..         ...     ...           ...  
95    La Jolla  Calif.      250000.0  
96    New York    N.Y.       87070.0  
97  Washington    D.C.      226226.0  
98    New York    N.Y.      171825.0  
99   Las Vegas    Nev.      208276.0  

[100 rows x 7 columns]

## Modificar DataFrames

In [11]:
#Renombrar la columna Sales a Sales_in_millions del dataframe future50
future50 = future50.rename(columns={'Sales':'Sales_in_millions'})
future50

Rank                        Restaurant                  Location  \
0      1                        Evergreens            Seattle, Wash.   
1      2                       Clean Juice           Charlotte, N.C.   
2      3                          Slapfish  Huntington Beach, Calif.   
3      4                        Clean Eatz          Wilmington, N.C.   
4      5                         Pokeworks            Irvine, Calif.   
5      6                       Playa Bowls             Belmar,  N.J.   
6      7                  The Simple Greek            Blue Bell, Pa.   
7      8                         Melt Shop            New York, N.Y.   
8      9                        Creamistry      Yorba Linda,  Calif.   
9     10              Joella's Hot Chicken           Louisville, Ky.   
10    11                     Eggs Up Grill         Spartanburg, S.C.   
11    12                          Dog Haus          Pasadena, Calif.   
12    13                  Teriyaki Madness             Denver, Colo.   
13    14                    Bluestone Lane            New York, N.Y.   
14    15                 Original ChopShop              Plano, Texas   
15    16                 Rapid Fired Pizza           Kettering, Ohio   
16    17           Ike's Love & Sandwiches     San Francisco, Calif.   
17    18                    Vitality Bowls         San Ramon, Calif.   
18    19         Hawkers Asian Street Fare             Orlando, Fla.   
19    20          Maple Street Biscuit Co.         Orange Park, Fla.   
20    21                   Bulla Gastrobar               Doral, Fla.   
21    22                       Duck Donuts        Mechanicsburg, Pa.   
22    23                   The Little Beet            New York, N.Y.   
23    24                   Joe & The Juice            New York, N.Y.   
24    25                          By Chloe            New York, N.Y.   
25    26              Sugarfire Smokehouse             Olivette, Mo.   
26    27               Bibibop Asian Grill            Columbus, Ohio   
27    28               Bubbakoo's Burritos       Wall Township, N.J.   
28    29                Dos Toros Taqueria            New York, N.Y.   
29    30                   The Pizza Press           Anaheim, Calif.   
30    31                    The Lost Cajun             Frisco, Texas   
31    32                        Just Salad            New York, N.Y.   
32    33        Jeni's Splendid Ice Creams            Columbus, Ohio   
33    34                          Boqueria            New York, N.Y.   
34    35         Quickway Japanese Hibachi              Fairfax, Va.   
35    36                         Surcheros              Douglas, Ga.   
36    37                   Famous Toastery           Charlotte, N.C.   
37    38                  Culinary Dropout         Scottsdale, Ariz.   
38    39                     Condado Tacos            Columbus, Ohio   
39    40           The Flying Biscuit Cafe              Atlanta, Ga.   
40    41             Blue Sushi Sake Grill               Omaha, Neb.   
41    42                    The Human Bean             Medford, Ore.   
42    43   Spitz Mediterranean Street Food       Los Angeles, Calif.   
43    44                      Tacos 4 Life              Conway, Ark.   
44    45    Pita Mediterranean Street Food             Fairburn, Ga.   
45    46                       LA Crawfish            McAllen, Texas   
46    47                            &pizza          Washington, D.C.   
47    48               Super Duper Burgers     San Francisco, Calif.   
48    49                   StoneFire Grill      Agoura Hills, Calif.   
49    50  Gus's World Famous Fried Chicken            Memphis, Tenn.   

    Sales_in_millions YOY_Sales  Units YOY_Units  Unit_Volume Franchising  
0                  24    130.5%     26    116.7%         1150          No  
1                  44    121.9%    105     94.4%          560         Yes  
2                  21     81.0%     21     90.9%         1370         Yes  
3                  25     7

In [12]:
#Renombrar la columna Sales a Sales_in_millions de top250
top250 = top250.rename(columns={'Sales':'Sales_in_millions'})
top250

Rank           Restaurant  \
0       1           McDonald's   
1       2            Starbucks   
2       3          Chick-fil-A   
3       4            Taco Bell   
4       5          Burger King   
..    ...                  ...   
245   246      Rainforest Cafe   
246   247                  PDQ   
247   248        Lupe Tortilla   
248   249  Cook-Out Restaurant   
249   250             Jollibee   

                                               Content  Sales_in_millions  \
0                                                  NaN              40412   
1                                                  NaN              21380   
2    While Popeyes got a lot of the chicken buzz in...              11320   
3                                                  NaN              11293   
4                                                  NaN              10204   
..                                                 ...                ...   
245                                                NaN                129   
246                                                NaN                127   
247                                                NaN                127   
248                                                NaN                126   
249                                                NaN                126   

    YOY_Sales  Units YOY_Units Headquarters             Segment_Category  
0        4.9%  13846     -0.5%          NaN       Quick Service & Burger  
1        8.6%  15049      3.0%          NaN  Quick Service & Coffee Cafe  
2       13.0%   2470      5.0%          NaN      Quick Service & Chicken  
3        9.0%   6766      2.7%          NaN      Quick Service & Mexican  
4        2.7%   7346      0.2%          NaN       Quick Service & Burger  
..        ...    ...       ...          ...                          ...  
245    -10.4%     18     -5.3%          NaN                  Varied Menu  
246     -5.5%     56    -11.1%          NaN                      Chicken  
247     12.1%     25      8.7%          NaN                      Mexican  
248     10.1%    270      7.1%          NaN                       Burger  
249     15.2%     40     11.1%          NaN                      Chicken  

[250 rows x 9 columns]

In [13]:
#Eliminar las columnas Content y Headquarters de top250 --> ["Content", "Headquarters"]
top250 = top250.drop(columns=['Content','Headquarters'])
top250

Rank           Restaurant  Sales_in_millions YOY_Sales  Units YOY_Units  \
0       1           McDonald's              40412      4.9%  13846     -0.5%   
1       2            Starbucks              21380      8.6%  15049      3.0%   
2       3          Chick-fil-A              11320     13.0%   2470      5.0%   
3       4            Taco Bell              11293      9.0%   6766      2.7%   
4       5          Burger King              10204      2.7%   7346      0.2%   
..    ...                  ...                ...       ...    ...       ...   
245   246      Rainforest Cafe                129    -10.4%     18     -5.3%   
246   247                  PDQ                127     -5.5%     56    -11.1%   
247   248        Lupe Tortilla                127     12.1%     25      8.7%   
248   249  Cook-Out Restaurant                126     10.1%    270      7.1%   
249   250             Jollibee                126     15.2%     40     11.1%   

                Segment_Category  
0         Quick Service & Burger  
1    Quick Service & Coffee Cafe  
2        Quick Service & Chicken  
3        Quick Service & Mexican  
4         Quick Service & Burger  
..                           ...  
245                  Varied Menu  
246                      Chicken  
247                      Mexican  
248                       Burger  
249                      Chicken  

[250 rows x 7 columns]

In [21]:
#Hacer que las unidades de la columna sales coincidan en las tablas
#Pasos:
# Dividir la columna del dataframe Sales entre 1000000
sales_mils = ind100['Sales']/1000000
#Asignar a la columna del dataframe Sales la lista sales_mils
ind100["Sales"] = sales_mils
#Renombrar la columna Sales a Sales_in_millions
ind100 = ind100.rename(columns={'Sales':'Sales_in_millions'})
ind100

Rank                           Restaurant  Sales_in_millions  \
0      1             Carmine's (Times Square)          39.080335   
1      2                The Boathouse Orlando          35.218364   
2      3                     Old Ebbitt Grill          29.104017   
3      4  LAVO Italian Restaurant & Nightclub          26.916180   
4      5             Bryant Park Grill & Cafe          26.900000   
..   ...                                  ...                ...   
95    96                 George's at the Cove          12.194000   
96    97                            Le Coucou          12.187523   
97    98                              Mi Vida          12.032014   
98    99                               Upland          11.965564   
99   100               Virgil's Real Barbecue          11.391678   

    Average Check        City   State  Meals Served  
0              40    New York    N.Y.      469803.0  
1              43    Orlando     Fla.      820819.0  
2              33  Washington    D.C.      892830.0  
3              90    New York    N.Y.      198500.0  
4              62    New York    N.Y.      403000.0  
..            ...         ...     ...           ...  
95             80    La Jolla  Calif.      250000.0  
96             95    New York    N.Y.       87070.0  
97             38  Washington    D.C.      226226.0  
98             52    New York    N.Y.      171825.0  
99             27   Las Vegas    Nev.      208276.0  

[100 rows x 7 columns]

### Agrupar Independent 100 por Restaurant

In [22]:
# Agrupar las filas por Restaurant y se suman
ind100_grouped = ind100.groupby(['Restaurant']).sum()
# Eliminar las columnas Rank y Average Check
ind100_grouped = ind100_grouped.drop(columns=['Rank', 'Average Check'])
# Agrupar por restaurant y average check, después añadir a ind100_grouped
avg_check = ind100.groupby(['Restaurant']).mean()
ind100_grouped['Average Check'] = avg_check['Average Check']
# Agrupar por restaurant y obtener el número de veces que aparece ese restaurante
counts = ind100.groupby(['Restaurant']).count()
ind100_grouped['Restaurant Count'] = counts['Rank']
# Ordenar por Sales de manera descendente
ind100_grouped = ind100_grouped.sort_values(by='Sales_in_millions', ascending=False)
# Restablecer el índice para mostrar el nuevo rango
ind100_grouped = ind100_grouped.reset_index()
# Mostrar cómo quedaría el DataFrame
ind100_grouped.head()

Restaurant  Sales_in_millions  Meals Served  \
0  Joe's Seafood, Prime Steak & Stone Crab          69.501000      816350.0   
1                 Gibsons Bar & Steakhouse          63.724826      877430.0   
2                 Carmine's (Times Square)          39.080335      469803.0   
3                    The Boathouse Orlando          35.218364      820819.0   
4                         Old Ebbitt Grill          29.104017      892830.0   

   Average Check  Restaurant Count  
0           86.0                 3  
1           80.0                 3  
2           40.0                 1  
3           43.0                 1  
4           33.0                 1

### Juntar Future50 e Ind100 por localización

In [26]:
# Paso 1. Hacer una copia de Future 50 df en la variable future50_df
future50_df = future50.copy()
future50.head()

Rank   Restaurant                  Location  Sales_in_millions YOY_Sales  \
0     1   Evergreens            Seattle, Wash.                 24    130.5%   
1     2  Clean Juice           Charlotte, N.C.                 44    121.9%   
2     3     Slapfish  Huntington Beach, Calif.                 21     81.0%   
3     4   Clean Eatz          Wilmington, N.C.                 25     79.7%   
4     5    Pokeworks            Irvine, Calif.                 49     77.1%   

   Units YOY_Units  Unit_Volume Franchising  
0     26    116.7%         1150          No  
1    105     94.4%          560         Yes  
2     21     90.9%         1370         Yes  
3     46     58.6%          685         Yes  
4     50     56.3%         1210         Yes

In [27]:
# Paso 2. Dividimos la columna Location de Future50 en dos nuevas columnas, City y State en un nuevo DataFrame
# Para ello usamos el método split y le pedimos que divida por comas
future50_split = future50_df["Location"].str.split(",", n = 1, expand = True)
future50_split.head()

0        1
0           Seattle    Wash.
1         Charlotte     N.C.
2  Huntington Beach   Calif.
3        Wilmington     N.C.
4            Irvine   Calif.

In [28]:
# Paso 3. Crear las columnas City y State dentro de future50_df 
# Pista: Se instancia el valor como en las listas (con el índice)
future50_df["City"]= future50_split[0]
future50_df["State"]= future50_split[1]
# Eliminar las columnas innecesarias --> ['Location', 'YOY_Sales', 'Units', 'YOY_Units', 'Unit_Volume', 'Franchising' ]
future50_df = future50_df.drop(columns=['Location','YOY_Sales', 'Units', 'YOY_Units', 'Unit_Volume', 'Franchising'])
future50_df.head()

Rank   Restaurant  Sales_in_millions              City    State
0     1   Evergreens                 24           Seattle    Wash.
1     2  Clean Juice                 44         Charlotte     N.C.
2     3     Slapfish                 21  Huntington Beach   Calif.
3     4   Clean Eatz                 25        Wilmington     N.C.
4     5    Pokeworks                 49            Irvine   Calif.

In [32]:
# Paso 4. Crear una copia de Ind100 df en la variable ind100_df 
ind100_df = ind100.copy()
#Ahora elimina columnas innecesarias --> ['Average Check', 'Meals Served']
ind100_df = ind100_df.drop(columns=['Average Check', 'Meals Served'])
ind100_df.head()

Rank                           Restaurant  Sales_in_millions        City  \
0     1             Carmine's (Times Square)          39.080335    New York   
1     2                The Boathouse Orlando          35.218364    Orlando    
2     3                     Old Ebbitt Grill          29.104017  Washington   
3     4  LAVO Italian Restaurant & Nightclub          26.916180    New York   
4     5             Bryant Park Grill & Cafe          26.900000    New York   

  State  
0  N.Y.  
1  Fla.  
2  D.C.  
3  N.Y.  
4  N.Y.

In [35]:
# Paso 5. Combinar el dataframe Future50 y el dataframe Ind100. Guardar el resultado en la variable top150_df
top150_df = pd.concat([future50, ind100], axis=0)
top150_df

Rank              Restaurant                  Location  Sales_in_millions  \
0      1              Evergreens            Seattle, Wash.          24.000000   
1      2             Clean Juice           Charlotte, N.C.          44.000000   
2      3                Slapfish  Huntington Beach, Calif.          21.000000   
3      4              Clean Eatz          Wilmington, N.C.          25.000000   
4      5               Pokeworks            Irvine, Calif.          49.000000   
..   ...                     ...                       ...                ...   
95    96    George's at the Cove                       NaN          12.194000   
96    97               Le Coucou                       NaN          12.187523   
97    98                 Mi Vida                       NaN          12.032014   
98    99                  Upland                       NaN          11.965564   
99   100  Virgil's Real Barbecue                       NaN          11.391678   

   YOY_Sales  Units YOY_Units  Unit_Volume Franchising  Average Check  \
0     130.5%   26.0    116.7%       1150.0          No            NaN   
1     121.9%  105.0     94.4%        560.0         Yes            NaN   
2      81.0%   21.0     90.9%       1370.0         Yes            NaN   
3      79.7%   46.0     58.6%        685.0         Yes            NaN   
4      77.1%   50.0     56.3%       1210.0         Yes            NaN   
..       ...    ...       ...          ...         ...            ...   
95       NaN    NaN       NaN          NaN         NaN           80.0   
96       NaN    NaN       NaN          NaN         NaN           95.0   
97       NaN    NaN       NaN          NaN         NaN           38.0   
98       NaN    NaN       NaN          NaN         NaN           52.0   
99       NaN    NaN       NaN          NaN         NaN           27.0   

          City   State  Meals Served  
0          NaN     NaN           NaN  
1          NaN     NaN           NaN  
2          NaN     NaN           NaN  
3          NaN     NaN           NaN  
4          NaN     NaN           NaN  
..         ...     ...           ...  
95    La Jolla  Calif.      250000.0  
96    New York    N.Y.       87070.0  
97  Washington    D.C.      226226.0  
98    New York    N.Y.      171825.0  
99   Las Vegas    Nev.      208276.0  

[150 rows x 13 columns]

In [36]:
# Paso 6. Crear una nueva columna que clasifica según la ubicación
# Ordenar por Sales de manera descendente
top150_df = top150_df.sort_values(by='Sales_in_millions', ascending=False)
# Restablecer el índice para mostrar el nuevo rango para location df
top150_df = top150_df.reset_index()
# Hacer que el índice empiece en 1
top150_df.index = np.arange(1, len(top150_df) + 1)
# Renombrar el índice de Rank para mostrar el nuevo rango
top150_df.index.names = ['Rank']
# Eliminar columnas innecesarias
top150_df = top150_df.drop(columns = ['index', 'Rank'])
top150_df


Restaurant        Location  Sales_in_millions YOY_Sales  \
Rank                                                                        
1                  Pokeworks  Irvine, Calif.          49.000000     77.1%   
2      Blue Sushi Sake Grill     Omaha, Neb.          49.000000     19.5%   
3             Bluestone Lane  New York, N.Y.          48.000000     33.0%   
4                LA Crawfish  McAllen, Texas          48.000000     17.6%   
5             The Human Bean   Medford, Ore.          47.000000     19.0%   
...                      ...             ...                ...       ...   
146     George's at the Cove             NaN          12.194000       NaN   
147                Le Coucou             NaN          12.187523       NaN   
148                  Mi Vida             NaN          12.032014       NaN   
149                   Upland             NaN          11.965564       NaN   
150   Virgil's Real Barbecue             NaN          11.391678       NaN   

      Units YOY_Units  Unit_Volume Franchising  Average Check        City  \
Rank                                                                        
1      50.0     56.3%       1210.0         Yes            NaN         NaN   
2      14.0     16.7%       3500.0          No            NaN         NaN   
3      48.0     37.1%       1175.0          No            NaN         NaN   
4      25.0     13.6%       2050.0         Yes            NaN         NaN   
5      97.0     19.8%        535.0         Yes            NaN         NaN   
...     ...       ...          ...         ...            ...         ...   
146     NaN       NaN          NaN         NaN           80.0    La Jolla   
147     NaN       NaN          NaN         NaN           95.0    New York   
148     NaN       NaN          NaN         NaN           38.0  Washington   
149     NaN       NaN          NaN         NaN           52.0    New York   
150     NaN       NaN          NaN         NaN           27.0   Las Vegas   

       State  Meals Served  
Rank                        
1        NaN           NaN  
2        NaN           NaN  
3        NaN           NaN  
4        NaN           NaN  
5        NaN           NaN  
...      ...           ...  
146   Calif.      250000.0  
147     N.Y.       87070.0  
148     D.C.      226226.0  
149     N.Y.      171825.0  
150     Nev.      208276.0  

[150 rows x 12 columns]

## Cargar DataFrames en la Base de Datos

In [37]:
#Rellena la siguiente variable con tu nombre y apellido (Todo junto y sin espacios)
nombreBD = 'ISI'
linkConexion = 'mysql+mysqlconnector://alumno:385bd4cf713a8a36@52.47.172.110/' + nombreBD
#Establecemos la conexión
engine = create_engine(linkConexion)

In [38]:
if not database_exists(engine.url):
    create_database(engine.url)

In [39]:
# Sube el dataframes a sql para future50df.
# PISTA: Usa la funcion to_sql() de sqlalchemy
future50.to_sql('tablafuture50', con=engine)

In [40]:
# Sube el dataframes a sql para ind100 df.
ind100.to_sql('tablaind100', con=engine)

In [41]:
# Sube el dataframes a sql para top250 df.
top250.to_sql('tablatop250', con=engine)

In [43]:
# Sube el dataframes a sql para top150 df.
top150_df.to_sql('tablatop150', con=engine)

### Confirmar que los datos han sido añadidos consultando la tabla Top150

In [44]:
# Haz un select de la tabla top150
query = " select * from tablatop150"
df  = pd.read_sql(query, con = engine )
df

Rank              Restaurant        Location  Sales_in_millions  \
0       1               Pokeworks  Irvine, Calif.          49.000000   
1       2   Blue Sushi Sake Grill     Omaha, Neb.          49.000000   
2       3          Bluestone Lane  New York, N.Y.          48.000000   
3       4             LA Crawfish  McAllen, Texas          48.000000   
4       5          The Human Bean   Medford, Ore.          47.000000   
..    ...                     ...             ...                ...   
145   146    George's at the Cove            None          12.194000   
146   147               Le Coucou            None          12.187523   
147   148                 Mi Vida            None          12.032014   
148   149                  Upland            None          11.965564   
149   150  Virgil's Real Barbecue            None          11.391678   

    YOY_Sales  Units YOY_Units  Unit_Volume Franchising  Average Check  \
0       77.1%   50.0     56.3%       1210.0         Yes            NaN   
1       19.5%   14.0     16.7%       3500.0          No            NaN   
2       33.0%   48.0     37.1%       1175.0          No            NaN   
3       17.6%   25.0     13.6%       2050.0         Yes            NaN   
4       19.0%   97.0     19.8%        535.0         Yes            NaN   
..        ...    ...       ...          ...         ...            ...   
145      None    NaN      None          NaN        None           80.0   
146      None    NaN      None          NaN        None           95.0   
147      None    NaN      None          NaN        None           38.0   
148      None    NaN      None          NaN        None           52.0   
149      None    NaN      None          NaN        None           27.0   

           City   State  Meals Served  
0          None    None           NaN  
1          None    None           NaN  
2          None    None           NaN  
3          None    None           NaN  
4          None    None           NaN  
..          ...     ...           ...  
145    La Jolla  Calif.      250000.0  
146    New York    N.Y.       87070.0  
147  Washington    D.C.      226226.0  
148    New York    N.Y.      171825.0  
149   Las Vegas    Nev.      208276.0  

[150 rows x 13 columns]

In [47]:
# Recupera los restaurantes "Virgil's Real Barbecue" de la tabla top150
query = '''select * from tablatop150 where restaurant = "Virgil's Real Barbecue" '''
df  = pd.read_sql(query, con = engine )
df

Rank              Restaurant Location  Sales_in_millions YOY_Sales Units  \
0   143  Virgil's Real Barbecue     None          12.245998      None  None   
1   150  Virgil's Real Barbecue     None          11.391678      None  None   

  YOY_Units Unit_Volume Franchising  Average Check       City State  \
0      None        None        None           31.0   New York  N.Y.   
1      None        None        None           27.0  Las Vegas  Nev.   

   Meals Served  
0      251800.0  
1      208276.0

### Resumen

A partir de una fuente de datos opensource en [Kaggle](https://www.kaggle.com/datasets/michau96/restaurant-business-rankings-2020) hemos creado montado ETL. Extraído, procesado y cargado los datos en una BBDD (base de datos).


---


## Obtención de datos

INPUT: 3 archivos CSV con las clasificaciones de los mejores restaurantes de 2020.



*   Future 50:
Clasificaciones de los futuros 50 mejores restaurantes.
*   Independence 100:
Clasificaciones de los 100 mejores restaurantes independientes de 2020.
*   Top 250:
Clasificaciones de los 250 mejores restaurantes de 2020.

Todos los datos se extrajeron de Restaurant Business Online y están basados en informes financieros de la empresa, encuestas, documentos de las franquicias y algoritmos de valoración.

## Transformación

1. Cambiamos el nombre a todas las columnas Sales para indicar que las unidades estaban en millones y corregir el dataframe Independence 100 cuya columna Sales no representamos en millones para que hubiera consistencia en las tablas.
2. Eliminamos algunas columnas adicionales que no vamos a usar en futuros análisis. Seleccionamos "Content" y "Headquarter" del dataframe top 250 ya que contenían muchos valores NaN y no proporcionaban información útil.
3. Agrupamos el dataframe independent 100 por Restaurant para combinar cualquier duplicado dentro de la tabla. Sumamos las columnas "Sales in Millions" y "Meals Served", calculamos el promedio de la columna "Average Check" y hacemos un recuento de cada restaurante para que se pueda ver si estaba representado en la tabla original varias veces. Finalmente, ordenamos este nuevo dataframe por Sales in Millions en orden descendente para ver los nuevos restaurantes mejor clasificados.
4. Para mantener la integridad de nuestro dataframe Future 50, hacemos una copia. Nuestro objetivo es fusionar el dataframe con la lista Independent 100 en función de los detalles de la ubicación.
5. Usando el dataframe copiado, separamos los valores de "Location" en dos columnas separadas ("City" y "State") y los almacenamos en un nuevo dataframe.
6. Para agregar los datos de "City" y "State" a nuestra copia original del dataframe Future 50, creamos las nuevas columnas y luego insertamos los valores separados en el paso anterior. Finalmente, eliminamos las columnas innecesarias.
7. Para mantener la integridad de nuestro dataframe Independent 100, hacemos una copia y luego eliminamos las columnas innecesarias. Hacemos que los dataframes copiados de Future 50 e Independent 100 tengan las mismas columnas para combinarlos fácilmente. Finalmente concatenamos los dos dataframes para hacer una lista del top 150.
8. Cogemos el dataframe Top 150 y ordenamos las filas por Sales in Millions. A continuación, reiniciamos el índice y lo hacemos contar desde 1 para mostrar la nueva clasificación. Finalmente, cambiamos el nombre del índice a Rank y eliminamos las columnas innecesarias.

##Subida a la base de datos

Una vez habiendo hecho un procesado de los datos nos disponemos a hacer los inserts a la BBDD. 
1. Nos conectamos a la base de datos.
2. Y, por último, subimos nuestros dataframes para que se almacenen en tablas.



